[MRG] handle sparse x and intercept in _RidgeGCV

jeromedockes opened the pull request:
closes #2354, closes #13325, closes #5364, closes #13321, closes #4781. 

Currently, if the design matrix is sparse, no intercept is fitted by
`_RidgeGCV`, even if `fit_intercept` is true.

This is an attempt to handle this case correctly when `gcv_mode` is "eigen",
which uses an eigendecompositon of the Gram matrix.
(`gcv_mode="svd"` cannot handle sparse matrices because the left singular
vectors of X are dense and this could use too much memory).

This is work in progress and for now, to make comparison easier, suggested
changes are implemented in a subclass of `_RidgeGCV`, `_WIPNewRidgeGCV`.

jeromedockes committed:
handle sparse x and intercept in _RidgeGCV with gcv_mode='eigen'

jeromedockes commented:
some figures to compare the current and proposed behaviors, which can be
reproduced with:

https://gist.github.com/jeromedockes/7fca2698f6ab7413903c782becc22c59

![coef](https://user-images.githubusercontent.com/9196501/53639425-7d002d00-3c29-11e9-97e3-b5b81e850e1d.png)
![intercept](https://user-images.githubusercontent.com/9196501/53639428-7ffb1d80-3c29-11e9-8879-b95be4d683ea.png)

jeromedockes commented:
note: the proposed solution passes `check_estimator_sparse_dense` in #7590 while the current one does not

jeromedockes committed:
start svd

jeromedockes committed:
iter

jeromedockes committed:
iter

jeromedockes committed:
remove warning

jeromedockes committed:
iter

jeromedockes committed:
details

jeromedockes committed:
only keep first min(n,p) singular values

jeromedockes committed:
comment

jeromedockes committed:
remove _WIPRidgeCV class

jeromedockes committed:
pep8

jeromedockes committed:
_check_gcv_mode util

jeromedockes committed:
fall back to 10-fold when sample weights, gcv and sparse design

jeromedockes committed:
add test to compare results of GCV with gridsearch leave-one-out

jeromedockes committed:
test with and without normalize

jeromedockes committed:
Merge branch 'master' into ridge_gcv_sparse_x_intercept

jeromedockes committed:
add sparse svd to _test_ridge_loo

jeromedockes committed:
remove test_ridge_cv_sparse_svd

jeromedockes committed:
smaller test data for ridgecv to speed up test

jeromedockes committed:
add uninformative features

jeromedockes committed:
add tests with sample weights

jeromedockes committed:
pep8

jeromedockes committed:
Merge remote-tracking branch 'upstream/master' into ridge_gcv_sparse_x_intercept

jeromedockes committed:
pep8

jeromedockes committed:
test sample weights with bigger alpha

jeromedockes committed:
clip slice for indexing sparse matrices with scipy 0.17

jeromedockes committed:
ignore gridsearch deprecation warning

jeromedockes commented:
ping @agramfort I think you can do a first review

agramfort reviewed (approved):
did not spot issues by reading the code and tests (sparse vs dense) are good.

LGTM @jeromedockes 

can you just update what's new page? thanks

jnothman commented:
Would @thomasjpfan like to review this then?

thomasjpfan commented:
I’ll look into this PR.

jeromedockes commented:
Actually I realized there is still a small issue when using sample weights: the
mean that gets removed from X is the average weighted by the sample weights, and
then X (and Y) is rescaled by the square roots of the sample weights. this means
that the design matrix actually used for the generalized cross-validation
(reweighted X) is not centered, and the constant vector ([1, 1, ..., 1]) is not a
singular vector of X; hence the regularization doesn't get cancelled for the
intercept (see #6178). the singular vector created by centering X becomes the
vector containing the square roots of the weights. I will have more time to look
into this tomorrow and will provide more details.
If we consider that handling sample weights correcly is a different issue, I can
also systematically fall back to 5-fold CV when sample weights are provided and
try to address this in another PR. what do you think?

jeromedockes committed:
don't center in _pre_compute when X is dense

jeromedockes committed:
update whats_new

jeromedockes committed:
use the square roots of sample weights rather than a column of ones

to capture the intercept in ridge generalized cross-val when sample
weights are provided

jeromedockes committed:
compare to singular vector rather than check singular value

to find intercept direction

jeromedockes committed:
add some comments + sample weights in sparse primal

jeromedockes committed:
duplicated line

glemaitre commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +311,72 @@ def test_ridge_individual_penalties():
     assert_raises(ValueError, ridge.fit, X, y)
 
 
+def test_ridge_gcv_vs_k_fold():
```
This might simplify a bit the readibility

```python
@pytest.mark.parametrize('gcv_mode', ['svd', 'eigen'])
@pytest.mark.parametrize('X_constructor', [np.asarray, sp.csr_matrix])
@pytest.mark.parametrize('X_shape', [(11, 8), (11, 20)])
@pytest.mark.parametrize('fit_intercept', [True, False])
@pytest.mark.parametrize('normalize', [True, False])
def test_ridge_gcv_vs_k_fold(gcv_mode, X_constructor, X_shape, fit_intercept,
                             normalize):
    n_samples, n_features = X_shape
    X, y = make_regression(
        n_samples=n_samples, n_features=n_features, n_targets=3,
        random_state=0, shuffle=False, noise=30., n_informative=5
    )
    X += 30 * np.random.RandomState(0).randn(X.shape[1])

    alphas = [1e-3, .1, 1., 10., 1e3]
    loo_ridge = RidgeCV(cv=n_samples, fit_intercept=fit_intercept,
                        alphas=alphas, scoring='neg_mean_squared_error',
                        normalize=normalize)
    gcv_ridge = RidgeCV(gcv_mode=gcv_mode, fit_intercept=fit_intercept,
                        alphas=alphas, normalize=normalize)

    loo_ridge.fit(X, y)

    X_gcv = X_constructor(X)
    gcv_ridge.fit(X_gcv, y)

    assert gcv_ridge.alpha_ == pytest.approx(loo_ridge.alpha_)
    assert_allclose(gcv_ridge.coef_, loo_ridge.coef_, rtol=1e-4)
    assert_allclose(gcv_ridge.intercept_, loo_ridge.intercept_, rtol=1e-4)
```

jeromedockes replied:
> This might simplify a bit the readibility

thanks!

jeromedockes committed:
parametrize test_ridge_gcv_vs_k_fold

jeromedockes committed:
better helper function names

jeromedockes committed:
handle sample weights with sparse X with gcv_mode='svd'

jeromedockes committed:
handle sparse x and sample weights when gcv_mode='eigen'

jeromedockes committed:
fix test + test sample weights with sparse x

jeromedockes committed:
refactor gcv mode selection

jeromedockes committed:
parametrize sample_weights test + test for p > n and fit_intercept=0

jeromedockes committed:
pep8

glemaitre commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -385,9 +458,10 @@ def _test_ridge_loo(filter_):
+    if filter_ == DENSE_FILTER:
+        ridge_gcv.fit(filter_(X_diabetes), y_diabetes,
+                      sample_weight=np.ones(n_samples))
+        assert_equal(ridge_gcv.alpha_, alpha_)
```
```suggestion
        assert ridge_gcv.alpha_ == pytest.approx(alpha_)
```

glemaitre reviewed (changes requested):
A couple of things regarding the tests.

jeromedockes committed:
compare gcv+sample weights with groupkfold

jeromedockes committed:
address @glemaitre review

jeromedockes committed:
test with small noise

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +993,100 @@ def _diag_dot(self, D, B):
+        if not center:
+            X_m = np.zeros(X.shape[1], dtype=X.dtype)
+            return safe_sparse_dot(X, X.T, dense_output=True), X_m
+        n = X.shape[0]
```
```suggestion
        n_samples = X.shape[0]
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +993,100 @@ def _diag_dot(self, D, B):
+        n = X.shape[0]
+        X_w = self._sqrt_sw_matrix.dot(X)
+        X_m, _ = mean_variance_axis(X_w, axis=0)
+        X_m = X_m * n / self._weight_sum
```
```suggestion
        X_m = X_m * n_samples / self._weight_sum
```

glemaitre replied:
We could do with inplace operator `*=`

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +993,100 @@ def _diag_dot(self, D, B):
+        X_m = X_m * n / self._weight_sum
+        X_mX = self._sqrt_sw[:, None] * safe_sparse_dot(
+            X_m, X.T, dense_output=True)
+        X_mX_m = np.empty((X.shape[0], X.shape[0]), dtype=X.dtype)
```
```suggestion
        X_mX_m = np.empty((n_samples, n_samples), dtype=X.dtype)
```

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -1087,55 +1310,54 @@ def fit(self, X, y, sample_weight=None):
-            raise ValueError("alphas cannot be negative. "
-                             "Got {} containing some "
-                             "negative value instead.".format(self.alphas))
+        if np.any(self.alphas <= 0):
```
This check can be moved earlier, before calling `_pre_compute`.

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -1087,55 +1310,54 @@ def fit(self, X, y, sample_weight=None):
+            self._weight_sum = X.shape[0]
+        self._sqrt_sw_matrix = sparse.dia_matrix(
+            (self._sqrt_sw, 0), shape=(X.shape[0], X.shape[0]))
+        precomputed = _pre_compute(X, y)
```
This can be computed right before `for i, alpha in enumerate(self.alphas):`, and using `v, Q, QT_y` originally is a little more clearer.

jeromedockes replied:
thanks!
I used a tuple instead of the individual variable names because the intermediary
variables computed by `_precompute_svd_sparse` are different, so it would be
strange to give them the same names. I still kept this precompute /
errors_and_values_helper organization in order to avoid changing the structure
of the code too much. another option would be to have pre_compute return a
dictionary, or have errors_and_values call pre_compute?

thomasjpfan reviewed (commented):
First pass. Thanks for the thoughts on sample weights, I will go over the implementation next.

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +993,100 @@ def _diag_dot(self, D, B):
+            X_m, X.T, dense_output=True)
+        X_mX_m = np.empty((X.shape[0], X.shape[0]), dtype=X.dtype)
+        X_mX_m[:, :] = np.dot(X_m, X_m)
+        X_mX_m = X_mX_m * self._sqrt_sw
```
```suggestion
        X_mX_m *= self._sqrt_sw * self._sqrt_sw[:, None]
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +993,100 @@ def _diag_dot(self, D, B):
+        X_m = X_m * n / self._weight_sum
+        return safe_sparse_dot(
+            X.T, X, dense_output=True) - self._weight_sum * np.outer(
+                X_m, X_m), X_m
```
It might be a bit easier to parse visually

```python
        return (safe_sparse_dot(X.T, X, dense_output=True) - 
                self._weight_sum * np.outer(X_m, X_m),
                X_m)
```

jeromedockes commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +998,100 @@ def _diag_dot(self, D, B):
+        X.dot(X.T)
+        """
+        if not center:
+            X_m = np.zeros(X.shape[1], dtype=X.dtype)
```
if center is false it means that x is dense and has already been centered, or
that we are not fitting an intercept, so in subsequent computations we should
consider that the mean is 0

glemaitre replied:
Ah right this is done in the preprocessing. Could you put a small comment then.

jeromedockes replied:
done :)

jeromedockes committed:
address comments by @glemaitre and @thomasjpfan

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +997,125 @@ def _diag_dot(self, D, B):
+    def _compute_covariance(self, X, center=True):
+        """Computes centered Gram matrix.
+
+        Notes
```
You can put it as a description of the function without a `Notes` section

jeromedockes commented:
> Thanks for the thoughts on sample weights

to be clear, I believe the issue mentionned in https://github.com/scikit-learn/scikit-learn/pull/13350#issuecomment-487399001 is now solved

jeromedockes committed:
fix docstring

jeromedockes committed:
comments

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+            if self._with_sw:
+                cov[-1, -1] = self._weight_sum
+            else:
+                cov[-1, -1] = n
```
```suggestion
                cov[-1, -1] = n_samples
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+    def _pre_compute_svd_sparse(self, X, y):
+        n, p = X.shape
+        cov = np.empty((p + 1, p + 1))
+        cov[:-1, :-1], X_m = self._compute_covariance(X, self.fit_intercept)
```
```suggestion
        cov[:-1, :-1], self._X_offset = self._compute_covariance(X, self.fit_intercept)
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+        # remove eigenvalues and vectors in the null space of X^T.X
+        s = s[kernel_size:]
+        V = V[:, kernel_size:]
+        self._X_offset = X_m
```
you can assign directly from the output of the function

glemaitre replied:
```suggestion
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+            intercept_dim = np.isclose(
+                U, self._normalized_sqrt_sw[:, None]).all(axis=0)
+        else:
+            intercept_dim = np.var(U, 0) < 1.e-12
```
Minimum of `np.argmin(np.min(U, axis=0))`

glemaitre replied:
after IRL, `isclose` is `rtol=1e-5` by default

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+    def _errors_and_values_svd_helper_sparse_no_intercept(
+            self, alpha, y, s, V, X):
+        """compute loo values and dual coef when X is sparse"""
+        n, p = X.shape
```
```suggestion
        n_samples, n_features = X.shape
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+        n, p = X.shape
+        w = 1 / (s + alpha)
+        A = (V * w).dot(V.T)
+        Xm = self._X_offset
```
```suggestion
        X_mean = self._X_offset
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+        Xm = self._X_offset
+        AXy = A.dot(safe_sparse_dot(X.T, y, dense_output=True))
+        y_hat = safe_sparse_dot(X, AXy, dense_output=True)
+        hat_diag = self._sparse_multidot_diag(X, A, Xm)
```
```suggestion
        hat_diag = self._sparse_multidot_diag(X, A, X_mean)
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+        intercept.
+
+        """
+        n, p = X.shape
```
```suggestion
        n_samples, n_features = X.shape
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+        w = 1 / (s + alpha)
+        w[intercept_dim] = 1 / s[intercept_dim]
+        A = (V * w).dot(V.T)
+        Xm = self._X_offset
```
```suggestion
        X_mean = self._X_offset
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+
+        Xop = sparse.linalg.LinearOperator(
+            matvec=matvec, matmat=matmat, rmatvec=rmatvec,
+            shape=(X.shape[0], X.shape[1] + 1), dtype=X.dtype)
```
```suggestion
            shape=(n_samples, n_feautres + 1), dtype=X.dtype
        )
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +997,129 @@ def _diag_dot(self, D, B):
+
+    def _sparse_multidot_diag(self, X, A, Xm):
+        """
+        compute the diagonal of (X - Xm).dot(A).dot((X - Xm).T)
```
"""Compute the diagonal of (X - X_mean).dot(A).dot((X - X_mean).T)
without explicitly centering `X` when `X` is sparse.
"""

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +997,129 @@ def _diag_dot(self, D, B):
+            batch = slice(start, min(X.shape[0], start + batch_size), 1)
+            X_batch = np.empty(
+                (X[batch].shape[0], X.shape[1] + self.fit_intercept),
+                dtype=X.dtype)
```
```suggestion
                dtype=X.dtype
            )
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +997,129 @@ def _diag_dot(self, D, B):
+                self._weight_sum * np.outer(X_mean, X_mean),
+                X_mean)
+
+    def _sparse_multidot_diag(self, X, A, Xm):
```
```suggestion
    def _sparse_multidot_diag(self, X, A, X_mean):
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +997,129 @@ def _diag_dot(self, D, B):
+                (X[batch].shape[0], X.shape[1] + self.fit_intercept),
+                dtype=X.dtype)
+            if self.fit_intercept:
+                X_batch[:, :-1] = X[batch].A - Xm * scale[batch][:, None]
```
```suggestion
                X_batch[:, :-1] = X[batch].A - X_mean * scale[batch][:, None]
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,6 +997,129 @@ def _diag_dot(self, D, B):
+    def _set_intercept(self, X_offset, y_offset, X_scale):
+        # add the mean of X which was computed separately if X is sparse
+        if getattr(self, '_X_offset', None) is not None:
+            X_offset = X_offset + self._X_offset * X_scale
```
```suggestion
            X_offset += self._X_offset * X_scale
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+        def matvec(v):
+            return safe_sparse_dot(
+                X, v[:-1], dense_output=True
+            ) - sw * Xm.dot(v[:-1]) + v[-1] * sw
```
```suggestion
            ) - sw * X_mean.dot(v[:-1]) + v[-1] * sw
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+        def matmat(v):
+            return safe_sparse_dot(
+                X, v[:-1], dense_output=True
+            ) - sw[:, None] * Xm.dot(v[:-1]) + v[-1] * sw[:, None]
```
```suggestion
            ) - sw[:, None] * X_mean.dot(v[:-1]) + v[-1] * sw[:, None]
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+
+        def rmatvec(v):
+            v = v.ravel()
+            res = np.empty(X.shape[1] + 1)
```
```suggestion
            res = np.empty(n_features + 1)
```

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1023,26 +1157,136 @@ def _values(self, alpha, y, v, Q, QT_y):
+            v = v.ravel()
+            res = np.empty(X.shape[1] + 1)
+            res[:-1] = safe_sparse_dot(
+                X.T, v, dense_output=True) - Xm * sw.dot(v)
```
```suggestion
                X.T, v, dense_output=True) - X_mean * sw.dot(v)
```

glemaitre reviewed (commented):
Some other comments but it looks good.

glemaitre committed:
Apply suggestions from code review

Co-Authored-By: jeromedockes <jerome@dockes.org>

jeromedockes committed:
fix variable name

jeromedockes committed:
better names for _RidgeGCV private helper functions

jeromedockes committed:
compare singular vectors to +- sqrt sample weights

instead of checking variance

jeromedockes commented on sklearn/linear_model/ridge.py:
```diff
@@ -1087,55 +1348,49 @@ def fit(self, X, y, sample_weight=None):
+        gcv_mode = _check_gcv_mode(X, self.gcv_mode)
 
         if gcv_mode == 'eigen':
             _pre_compute = self._pre_compute
```
renamed it to `_decompose_gram` since it computes an eigenvalue decomposition of the gram matrix

jeromedockes commented on sklearn/linear_model/ridge.py:
```diff
@@ -1004,9 +1129,18 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+                intercept_dim = np.isclose(
+                    Q, self._normalized_sqrt_sw[:, None]).all(axis=0)
+            else:
+                intercept_dim = np.var(Q, 0) < 1.e-12
```
I changed it to compare to what the singular vector we are looking for is, ie +- normalized vector of ones, to be consistent to the case with sample weights.
it is indeed possible to look for the vector with the smallest angle rather than testing with `np.isclose`, do you think it is better?

jeromedockes replied:
I changed it to cancel regularization on the singular vector which has the smallest angle (modulo pi) with the direction of the intercept

jeromedockes committed:
docstring

jeromedockes committed:
find singular vect with smallest angle instead of comparing values

jeromedockes committed:
reduce rtol in test

jeromedockes committed:
increase rtol to 5e-2 in test

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,7 +1015,131 @@ def _diag_dot(self, D, B):
+        """ compute the diagonal of (X - X_mean).dot(A).dot((X - X_mean).T)
+        without explicitely centering X nor computing X.dot(A)
+        when X is sparse.
+        """
```
Even if this is a private function you can add the parameters and what is returned.

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -995,7 +1015,131 @@ def _diag_dot(self, D, B):
+            X_offset += self._X_offset * X_scale
+        super()._set_intercept(X_offset, y_offset, X_scale)
+
+    def _decompose_gram(self, X, y):
```
You can put a small docstring (single line) stating for which case we should use it `n_samples` < `n_features`

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1015,48 +1164,168 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
-            raise TypeError("SVD not supported for sparse matrices")
-        if centered_kernel:
-            X = np.hstack((X, np.ones((X.shape[0], 1))))
+    def _decompose_covariance_sparse(self, X, y):
```
You can put a small docstring (single line) stating for which case we should use it `n_samples` > `n_features`

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -1015,48 +1164,168 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+
+    def _errors_and_values_covariance_sparse_no_intercept(
+            self, alpha, y, s, V, X):
+        """compute loo values and dual coef when X is sparse"""
```
Could you change slightly the docstring of the `_errors_and_values_...` functions such that they are all consistent. This is just a nitpick but like this we will have the same info in all of them.

glemaitre reviewed (commented):
The last nitpick regarding the code. I will try to see if I have some additional tests to add.

glemaitre commented on sklearn/linear_model/ridge.py:
```diff
@@ -921,6 +919,40 @@ def classes_(self):
+def _check_gcv_mode(X, gcv_mode):
+    possible_gcv_modes = {None, 'auto', 'svd', 'eigen'}
+    if gcv_mode not in possible_gcv_modes:
+        raise ValueError(
```
Apparently, we don't have a test for this case

jeromedockes replied:
thanks!

jeromedockes committed:
add test for _check_gcv_mode

jeromedockes committed:
remove temp variables after fitting ridge gcv

ogrisel commented on sklearn/linear_model/ridge.py:
```diff
@@ -1217,9 +1487,14 @@ def fit(self, X, y, sample_weight=None):
+                              parameters, cv=cv, scoring=self.scoring)
+            # note: unlike when using gcv, sample weights won't be used
+            # to compute the validation score so selected hyperparameter
+            # may differ
```
Maybe you could put this note in the docstring, for instance at the end of the paragraph that documents the `sample_weights` parameter.

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +311,96 @@ def test_ridge_individual_penalties():
+@pytest.mark.parametrize('fit_intercept', [True, False])
+@pytest.mark.parametrize('normalize', [True, False])
+@pytest.mark.parametrize('noise', [1., 30.])
+def test_ridge_gcv_vs_k_fold(gcv_mode, X_constructor, X_shape, fit_intercept,
```
Because this is not k-fold CV, may I suggest instead:

```suggestion
def test_ridge_gcv_vs_ridge_loo_cv(gcv_mode, X_constructor, X_shape, fit_intercept,
```

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +311,96 @@ def test_ridge_individual_penalties():
+
+    assert gcv_ridge.alpha_ == pytest.approx(loo_ridge.alpha_)
+    assert_allclose(gcv_ridge.coef_, loo_ridge.coef_, rtol=1e-4)
+    assert_allclose(gcv_ridge.intercept_, loo_ridge.intercept_, rtol=1e-4)
```
Very good test!

ogrisel replied:
I believe that there is [no example](https://scikit-learn.org/dev/modules/generated/sklearn.linear_model.RidgeCV.html#examples-using-sklearn-linear-model-ridgecv) that highlights the use and speed of GCV. I think it would be great to write one to show the equivalence between such an instance of `loo_ridge` and `gcv_ridge` on a small regression dataset such as Boston for instance.

One could plot the information collected with `store_cv_values=True`.

Finally one could also compare to the result obtained with 10-fold cross-validation (on neg MSE too).

jeromedockes committed:
more tests

ogrisel reviewed (commented):
Here is a first pass. I find the code a bit hard too read (but this is not your fault, the existing variable names are not very explicit) so I will do this review in several passes.

jeromedockes committed:
merge tests for gcv store_cv_values

jeromedockes commented on sklearn/linear_model/ridge.py:
```diff
@@ -1087,55 +1361,49 @@ def fit(self, X, y, sample_weight=None):
+        self._normalized_sqrt_sw = self._sqrt_sw / np.linalg.norm(
+            self._sqrt_sw)
+        self._sqrt_sw_matrix = sparse.dia_matrix(
+            (self._sqrt_sw, 0), shape=(X.shape[0], X.shape[0]))
```
I had stored them to avoid having long function signatures, at the moment they
are being removed at the end of the fit. we could also put the rest of fit in a
try/except block to make sure they get cleaned up even if `fit` fails.

but passing them as arguments definitely makes sense, I will make the change
right away.

ogrisel replied:
Yes I don't like the statefulness and implicitness of passing those as attributes + cleanup.

jeromedockes replied:
almost done, still need to get rid of `_X_offset`

ogrisel committed:
Update sklearn/linear_model/tests/test_ridge.py

Co-Authored-By: jeromedockes <jerome@dockes.org>

jeromedockes committed:
pep8

jeromedockes commented:
by the way, to check memory usage I will be adding more cases to the RidgeCV
benchmark in

https://github.com/jeromedockes/scikit-learn_benchmarks/tree/ridge_benchmark

glemaitre committed:
Update sklearn/linear_model/ridge.py

Co-Authored-By: jeromedockes <jerome@dockes.org>

glemaitre committed:
Apply suggestions from code review

Co-Authored-By: jeromedockes <jerome@dockes.org>

jeromedockes committed:
docstring

glemaitre commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,108 @@ def test_ridge_individual_penalties():
+    assert_allclose(gcv_ridge.intercept_, kfold.intercept_, rtol=5e-2)
+
+
+def test_check_gcv_mode():
```
you can change the test as:

```python
@pytest.mark.parametrize('mode', [True, 1, 5, 'bad', 'gcv', np.arange(3)])
def test_check_gcv_mode_error(mode):
    X, y = make_regression(n_samples=5, n_features=2)
    gcv = RidgeCV(gcv_mode=mode)
    with pytest.raises(ValueError, match="Unknown value for 'gcv_mode'"):
        gcv.fit(X, y)
        _check_gcv_mode(X, mode)


@pytest.mark.parametrize(
    'mode, mode_samples_sup_features, mode_features_sup_samples',
    [(None, 'svd', 'eigen'),
     ('auto', 'svd', 'eigen'),
     ('eigen', 'eigen', 'eigen'),
     ('svd', 'svd', 'svd')]
)
def test_check_gcv_mode_choice(mode, mode_samples_sup_features,
                               mode_features_sup_samples):
    X, _ = make_regression(n_samples=5, n_features=2)

    assert _check_gcv_mode(X, mode) == mode_samples_sup_features
    assert _check_gcv_mode(X.T, mode) == mode_features_sup_samples
```

jeromedockes replied:
thanks!


> ```python
>     with pytest.raises(ValueError, match="Unknown value for 'gcv_mode'"):
>         gcv.fit(X, y)
>         _check_gcv_mode(X, mode)
> ```

I never used this context manager before; both lines here need to raise (ie `gcv.fit` and `_check_gcv_mode` should both fail). does the second line get executed if the first one fails?

glemaitre replied:
I supposed that yes but let me check to be sure. Otherwise we call twice the context manager.

glemaitre replied:
So you are right, we need to split the 2 lines

```python
@pytest.mark.parametrize('mode', [True, 1, 5, 'bad', 'gcv', np.arange(3)])
def test_check_gcv_mode_error(mode):
    X, y = make_regression(n_samples=5, n_features=2)
    gcv = RidgeCV(gcv_mode=mode)

    err_msg = "Unknown value for 'gcv_mode'"
    with pytest.raises(ValueError, match=err_msg):
        gcv.fit(X, y)
    with pytest.raises(ValueError, match=err_msg):
        _check_gcv_mode(X, mode)
```

ogrisel replied:
No the second line will not be called if the first raises an exception. To check both, you need to use the context manager twice.

jeromedockes committed:
doc

jeromedockes committed:
_errors_and_values docstrings

glemaitre commented:
If @ogrisel has some additional idea for tests, it seems robust for the moment.
@jeromedockes do you make a quick list of the issues which could be addressed following this PR?

jeromedockes committed:
note on use of sample weights for computing the cv score

jeromedockes committed:
parametrize check_gcv_mode test as suggested by @glemaitre

jeromedockes committed:
don't store temp variables computed in ridgegcv fit in private attributes

jeromedockes committed:
don't store self._X_offset

jeromedockes committed:
improve test_check_gcv_mode_error

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -1087,55 +1386,43 @@ def fit(self, X, y, sample_weight=None):
         cv_values = np.zeros((n_samples * n_y, len(self.alphas)))
         C = []
 
         scorer = check_scoring(self, scoring=self.scoring, allow_none=True)
```
Nit: Moving the `check_scoring` before `n_y` to fail error, and have `C` and `cv_values` defined closer to where it will be used.

jeromedockes committed:
check_scoring earlier in _RidgeGCV fit

thomasjpfan commented:
BTW thanks for the refactor regarding the sample weights. It is a bit nicer now 👍

jeromedockes committed:
fix case where y.shape == (n_samples, 1)

jeromedockes commented:
> BTW thanks for the refactor regarding the sample weights. It is a bit nicer now +1

thanks!

jeromedockes commented:
I found an issue in the case where y is provided as a column vector and improved the tests to cover this case, sorry about that

jeromedockes committed:
more y shapes with sample weights

jeromedockes commented:
>  @jeromedockes do you make a quick list of the issues which could be addressed following this PR?

issues addressed by this PR:

- when X was sparse, no intercept was fitted (one of the failures of the test
  introduced in #13246 in response to #1572)
- when X was sparse, the fit was done using an eigendecomposition of X.X^T,
  which is a dense matrix larger than X itself when n_samples > n_features
  (#2354)
- when sample weights were provided, regularization wasn't cancelled on the
  intercept when computing the LOO values, i.e. the problem discussed in #6178
  wasn't addressed.
- when sample weights were provided, the fit was done using an
  eigendecomposition of X.X^T, which is inefficient and can use a lot of memory
  if n_samples > > n_features (#13321).

jeromedockes commented:
issues not addressed by this PR:

- this PR only concerns `RidgeCV`, when using gcv. `Ridge` is still listed as
  problematic in #13246

- it may be useful to avoid storing all the LOO errors when the user asks for
  `store_cv_values=False`, as this matrix can be large if there are several
  outputs and we only need the mean along one dimension (#13322)

jeromedockes commented:
note: this also passes the test added in #4490, once the test is made to take sample weights into account when computing the validation error, ie

```
def make_noisy_forward_data(
    n_samples=100,
    n_features=200,
    n_targets=10,
    train_frac=.8,
    noise_levels=None,
    random_state=42):
    """Creates a simple, dense, noisy forward linear model with multiple
    output."""
    rng = check_random_state(random_state)
    n_train = int(train_frac * n_samples)
    train = slice(None, n_train)
    test = slice(n_train, None)
    X = rng.randn(n_samples, n_features)
    W = rng.randn(n_features, n_targets)
    Y_clean = X.dot(W)
    if noise_levels is None:
        noise_levels = rng.randn(n_targets) ** 2
    noise_levels = np.atleast_1d(noise_levels) * np.ones(n_targets)
    noise = rng.randn(*Y_clean.shape) * noise_levels * Y_clean.std(0)
    Y = Y_clean + noise
    return X, Y, W, train, test


def test_ridge_gcv_with_sample_weights():
    from sklearn.model_selection import LeaveOneOut
    n_samples, n_features, n_targets = 20, 5, 7
    X, Y = datasets.make_regression(n_samples, n_features,
                                    n_targets=n_targets)
    alphas = np.logspace(-3, 3, 9)

    rng = np.random.RandomState(42)
    sample_weights = rng.randn(n_samples) ** 2
    cv = LeaveOneOut()
    cv_predictions = np.array([[
        Ridge(solver='cholesky', alpha=alpha, fit_intercept=False).fit(
            X[train], Y[train], sample_weight=sample_weights[train]
            ).predict(X[test])
        for train, test in cv.split(X, Y)] for alpha in alphas]).squeeze()

    cv_errors = Y[np.newaxis] - cv_predictions.reshape(
        len(alphas), n_samples, n_targets)

    ridge_gcv = _RidgeGCV(alphas=alphas, store_cv_values=True,
                          gcv_mode='eigen', fit_intercept=False)
    # emulate the sample weight stuff from _RidgeGCV
    ridge_gcv.fit(X, Y, sample_weight=sample_weights)
    loo_predictions = ridge_gcv.cv_values_

    assert_array_almost_equal(cv_errors ** 2 * sample_weights[None, :, None],
                              loo_predictions.transpose(2, 0, 1))
```

it also passes with sparse X and replacing 'eigen' with 'svd'

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,123 @@ def test_ridge_individual_penalties():
+@pytest.mark.parametrize('y_shape', [(11,), (11, 1), (11, 3)])
+@pytest.mark.parametrize('fit_intercept', [True, False])
+@pytest.mark.parametrize('normalize', [True, False])
+@pytest.mark.parametrize('noise', [1., 30.])
```
Doing the complete cartesian product of all those parameters is expensive: this sequence of tests take 24s on my laptop.

May I suggest to do the cartesian product only on a subset and hard code combinations for the remaining parameters:

```diff
diff --git a/sklearn/linear_model/tests/test_ridge.py b/sklearn/linear_model/tests/test_ridge.py
index 971066265..44358a398 100644
--- a/sklearn/linear_model/tests/test_ridge.py
+++ b/sklearn/linear_model/tests/test_ridge.py
@@ -315,10 +315,14 @@ def test_ridge_individual_penalties():
 @pytest.mark.parametrize('gcv_mode', ['svd', 'eigen'])
 @pytest.mark.parametrize('X_constructor', [np.asarray, sp.csr_matrix])
 @pytest.mark.parametrize('X_shape', [(11, 8), (11, 20)])
-@pytest.mark.parametrize('y_shape', [(11,), (11, 1), (11, 3)])
-@pytest.mark.parametrize('fit_intercept', [True, False])
-@pytest.mark.parametrize('normalize', [True, False])
-@pytest.mark.parametrize('noise', [1., 30.])
+@pytest.mark.parametrize(
+    'y_shape, fit_intercept, normalize, noise',
+    [
+        ((11,), True, True, 1.),
+        ((11, 1), True, False, 20.),
+        ((11, 3), False, False, 30.),
+    ]
+)
 def test_ridge_gcv_vs_ridge_loo_cv(
         gcv_mode, X_constructor, X_shape, y_shape,
         fit_intercept, normalize, noise):
```

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,123 @@ def test_ridge_individual_penalties():
+@pytest.mark.parametrize('fit_intercept', [True, False])
+@pytest.mark.parametrize('n_features', [11, 69])
+@pytest.mark.parametrize('y_shape', [(59,), (59, 1), (59, 3)])
+@pytest.mark.parametrize('noise', [1., 30.])
```
Same comment here: it believe we should only test for subset of hard-coded combinations so as to get this set of tests to run in less than 10s. The full Cartesian product takes 48s on my laptop.

jeromedockes replied:
both sequences now take 2s on my laptop

ogrisel replied:
Confirmed, thanks.

ogrisel reviewed (commented):
Some comments on test durations:

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,123 @@ def test_ridge_individual_penalties():
+     ('eigen', 'eigen', 'eigen'),
+     ('svd', 'svd', 'svd')]
+)
+def test_check_gcv_mode_choice(mode, mode_samples_sup_features,
```
`n_sup_p` feels french to me. Maybe use `n_greater_than_p` instead?

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,123 @@ def test_ridge_individual_penalties():
+     ('eigen', 'eigen', 'eigen'),
+     ('svd', 'svd', 'svd')]
+)
+def test_check_gcv_mode_choice(mode, mode_samples_sup_features,
```
Could you also parameterize this test with:

```python
@pytest.mark.parametrize("sparse", [True, False])
@pytest.mark.parametrize(
    'mode, mode_samples_sup_features, mode_features_sup_samples',
    [(None, 'svd', 'eigen'),
     ('auto', 'svd', 'eigen'),
     ('eigen', 'eigen', 'eigen'),
     ('svd', 'svd', 'svd')]
)
def test_check_gcv_mode_choice(sparse, mode, mode_samples_sup_features,
```
```

And check that that the results holds when `X` is converted to a CSR matrix?

ogrisel replied:
Speaking of which, the docstring for `gcv_mode` needs to be updated to take the handling of sparse data into account. Currently it reads:

```
gcv_mode : {None, 'auto', 'svd', eigen'}, optional
    Flag indicating which strategy to use when performing
    Generalized Cross-Validation. Options are::

        'auto' : use svd if n_samples > n_features or when X is a sparse
                 matrix, otherwise use eigen
        'svd' : force computation via singular value decomposition of X
                (does not work for sparse matrices)
        'eigen' : force computation via eigendecomposition of X^T X

    The 'auto' mode is the default and is intended to pick the cheaper
    option of the two depending upon the shape and format of the training
    data.
```

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,123 @@ def test_ridge_individual_penalties():
+                        alphas=alphas, scoring='neg_mean_squared_error',
+                        normalize=normalize)
+    gcv_ridge = RidgeCV(gcv_mode=gcv_mode, fit_intercept=fit_intercept,
+                        alphas=alphas, normalize=normalize)
```
Could you please update the dosctring of `RidgeCV` to explain what metric is used when `scoring=None` depending on the value of the `cv` parameter?

Currently it's not omitted:

```
scoring : string, callable or None, optional, default: None
    A string (see model evaluation documentation) or
    a scorer callable object / function with signature
    ``scorer(estimator, X, y)``.

cv : int, cross-validation generator or an iterable, optional
    Determines the cross-validation splitting strategy.
    Possible inputs for cv are:

    - None, to use the efficient Leave-One-Out cross-validation
    - integer, to specify the number of folds.
    - :term:`CV splitter`,
    - An iterable yielding (train, test) splits as arrays of indices.

    For integer/None inputs, if ``y`` is binary or multiclass,
    :class:`sklearn.model_selection.StratifiedKFold` is used, else,
    :class:`sklearn.model_selection.KFold` is used.

    Refer :ref:`User Guide <cross_validation>` for the various
    cross-validation strategies that can be used here.
```

Also when mentioning "efficient Leave-One-Out cross-validation' I think we should add "(also known as Generalized Cross Validation)".

jeromedockes committed:
reduce number of combinations in ridgegcv tests + smaller test data

ogrisel reviewed (commented):
This PR looks good, but still, here is another batch of reviews ;)

jeromedockes committed:
pep8

jeromedockes committed:
better parameter name

ogrisel commented:
I can work on an example if you wish.

jeromedockes committed:
changes suggested by @ogrisel

jeromedockes commented:
> I can work on an example if you wish.

as you prefer; I can also get started on an example. do you think it should be done in this PR or a separate one?

jeromedockes commented:
> I believe that there is no example that highlights the use and speed of GCV

I think it could also be nice to have somewhere in the documentation some suggestions as to when it is more efficient to use 10-fold cross-validation, for example for large datasets

ogrisel commented on sklearn/linear_model/ridge.py:
```diff
@@ -1015,48 +1182,176 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+            cov[-1] = 0
+            cov[:, -1] = 0
+            cov[-1, -1] = sqrt_sw.dot(sqrt_sw)
+        kernel_size = max(0, X.shape[1] - X.shape[0])
```
`kernel_size` is an ambiguous name in a machine learning context. Maybe `nullspace_dim` is more explicit IMO.

ogrisel commented:
> as you prefer; I can also get started on an example. do you think it should be done in this PR or a separate one?

We can keep it for another PR.

ogrisel commented:
> I think it could also be nice to have somewhere in the documentation some suggestions as to when it is more efficient to use 10-fold cross-validation, for example for large datasets

+1 for adding this in this current PR. There is a short section on Generalized CV in the `doc/modules/linear_model.rst` file.

jeromedockes committed:
kernel_size -> nullspace_dim

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,130 @@ def test_ridge_individual_penalties():
+        n_samples=11, n_features=n_features, n_targets=n_targets,
+        random_state=0, shuffle=False, noise=noise)
+    y = y.reshape(y_shape)
+    x += 30 * rng.randn(x.shape[1])
```
I don't understand this line. Is this to add some noise in the input space?

Also because `X` is a 2D array, please use capital case for this variable.

jeromedockes replied:
it is to add a constant to each column of  `X`, so that there is an intercept to fit (`X` produced by `make_regression` is centered). But it can be replaced with 
```
X += 30
```
(and the similar line in `test_ridge_gcv_sample_weights` too)

glemaitre replied:
Otherwise `X` would be centered and not a really good dataset to fit an intercept.

ogrisel replied:
Maybe also add a constant offset to y as well then :)

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +312,130 @@ def test_ridge_individual_penalties():
+    kfold = RidgeCV(
+        alphas=alphas, cv=splits, scoring='neg_mean_squared_error',
+        fit_intercept=fit_intercept)
+    with ignore_warnings():
```
Please add a comment to state which warning you want to ignore and why w do not care about this warning in this context.

jeromedockes committed:
specify why we ignore DeprecationWarning in ridgecv test

jeromedockes committed:
constant offset in test and x -> X

jeromedockes committed:
increase rtol to 1e-3

jeromedockes committed:
add constant offset to y

jeromedockes committed:
more doc on gcv

jeromedockes committed:
doctest skip

jeromedockes commented:
when the scorer is `None`, the stored `cv_values_` are the squared errors,
scaled by the sample weights.
otherwise, they are the leave-one-out predictions, but for the centered and
scaled data. thinking about a possible example, I wonder if it would be more
helpful to store the predicted values with added intercept and removing the
sample weight scaling?

jeromedockes committed:
perf: use LinearOperator transpose rather than _rmatvec

jeromedockes commented:
I made a small change: I defined the transpose of the `LinearOperator` in
`_X_operator` and `_Xt_operator`, to avoid computing a dot product with `y` by
looping over its columns in the multioutput case. I took the definition of the
operator out of the `_errors_and_values_covariance_sparse_intercept` so that it
doesn't become too long and to add a docstring and test for the linear operator.
sorry for making this small change late in the review process; only
`_Xt_operator._matmat` is new, the other methods are the ones which were defined
in `_errors_and_values_covariance_sparse_intercept` before.

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -147,7 +147,7 @@ as GridSearchCV except that it defaults to Generalized Cross-Validation
     >>> reg.alpha_                                          # doctest: +SKIP
     1.0
-    >>> reg.cv_values_
+    >>> reg.cv_values_                                      # doctest: +SKIP
```
You could use `# doctest: +NORMALIZE_WHITESPACE +ELLIPSIS`  instead and then use `...` after the third digit, e.g. `0.105...`.

jeromedockes committed:
don't skip ridge doctest in linear_model.rst

jeromedockes committed:
docstring

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -136,17 +136,27 @@ Setting the regularization parameter: generalized Cross-Validation
+efficient way to set the hyperparameter, and k-fold cross-validation can be
+faster. Specifying the value of the `cv` attribute will trigger the use of
+cross-validation with `GridSearchCV`, for example `cv=10` for 10-fold
+cross-validation.
```
Actually, on all the tests that I tried, Generalized CV with a non-trivial number of alphas (e.g. 5 or more) is always significantly faster than 5-fold cross-validation.

Do you have a case where GCV is not competitive with k-fold CV from a performance point of view?

jeromedockes replied:
Actually I am not able to find such a case, sorry for the noise. I'll fix the
documentation. I had in mind that in the medium p, large n case, gcv was not the
fastest method, and the discussion of the technical report cited in the doc
suggests that may be the case, but I couldn't observe it. This may be due to the
fact that the `RidgeCV` with `cv != None` uses a `GridSearch` and thus doesn't
reuse the svd decompositions or benefit from a warm start for other alphas
AFAIK.

ogrisel replied:
Indeed warm-starting and reused SVD might be interesting but should better be implemented in a separate PR.

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -136,17 +136,27 @@ Setting the regularization parameter: generalized Cross-Validation
+    >>> reg.cv_values_                                      # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    array([[0.105...   , 0.259...],
+           [0.022...   , 0.125...],
+           [0.902...   , 0.902...]])
```
Maybe we should not mention the `store_cv_values` / `cv_values_` things in the main documentation if we are not 100% satisfied with the current state of affairs (e.g. memory usage with large `n_samples` and handling of the sample weights). Let's keep this for another PR.

+1 for keeping `reg.alpha_` in the inline example though. I think you should even demonstrate a larger range for alpha, e.g. `alphas=np.logspace(-6, 6, 13)`.

ogrisel reviewed (commented):
Some comments about the documentation:

jeromedockes committed:
improve RidgeCV narrative doc

ogrisel commented on sklearn/linear_model/ridge.py:
```diff
@@ -1015,48 +1246,152 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+        # X already centered
+        X_mean = np.zeros(X.shape[1], dtype=X.dtype)
+        if self.fit_intercept:
+            intercept = sqrt_sw[:, None]
```
Because this is not the actual intercept value but a pseudo column that serves to emulate fitting an intercept, maybe it could be renamed to `intercept_column`.

ogrisel commented on sklearn/linear_model/ridge.py:
```diff
@@ -1015,48 +1246,152 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+            intercept = sqrt_sw[:, None]
+            X = np.hstack((X, intercept))
         # to emulate fit_intercept=True situation, add a column on ones
         # Note that by centering, the other columns are orthogonal to that one
```
typo: add a column **of** ones

Actually you add a column with the square root of the sample weights. Also this comments could be moved up right after the `if self.fit_intercept:` line.

ogrisel commented on sklearn/linear_model/ridge.py:
```diff
@@ -1015,48 +1246,152 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
+    def _values_covariance_dense(self, alpha, y, sqrt_sw, X_mean, v, U, UT_y):
+        G_diag, c = self._errors_and_values_covariance_dense(
+            alpha, y, sqrt_sw, v, U, UT_y)
         return y - (c / G_diag), c
```
This pattern is repeated many times. Wouldn't it be possible only call the `_errors_and_values` function and do the `squared_errors = (c / G_diag) ** 2` or `predictions = y - (c / G_diag)` transformations only once in the `_RidgeGCV.fit` method?

The name `_errors_and_values` is really misleading. It should probably be renamed to something more either more explicit more generic but less confusing such as `_solve`:

```python
G_diag, dual_coef = _solve(float(alpha), y, sqrt_sw, X_mean, *decomposition)

if error:
   squared_errors = (dual_coef / G_diag) ** 2
   ...
else:
   predictions = y - (dual_coef / G_diag)
   ...
```

ogrisel replied:
It would be great to find a better name for `G_diag` but I am not sure which.

jeromedockes replied:
maybe something like `inv_regularized_gram_diag`? one advandage of `G_diag` is that G is the notation used in the reference

ogrisel replied:
Alright. Let's keep `G_diag` then.

ogrisel reviewed (approved):
One last pass of review, hopefully this is the last. +1 for merge once the following comments are addressed:

ogrisel committed:
Update sklearn/linear_model/tests/test_ridge.py

Co-Authored-By: jeromedockes <jerome@dockes.org>

jeromedockes committed:
reduce tolerance to 1e-3 in test_ridge_gcv_sample_weights

jeromedockes committed:
duplicated line

jeromedockes commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -311,6 +313,151 @@ def test_ridge_individual_penalties():
+    assert_allclose(gcv_errors, kfold_errors, rtol=5e-2)
+    assert_allclose(gcv_errors, kfold_errors, rtol=5e-2)
+    assert_allclose(gcv_ridge.coef_, kfold.coef_, rtol=5e-2)
+    assert_allclose(gcv_ridge.intercept_, kfold.intercept_, rtol=5e-2)
```
1e-3 also works for me, I will try to push it and see if it fails on CI.


> Note that I also tried `1e-4` and it failed for only for the following two cases:

I'm not sure if it is due to chance but in both these cases we are forcing the use of 'svd' where 'auto' would pick using 'eigen' ( n > p )

ogrisel replied:
Yes I wouldn't mind too much. 1e-3 is probably good enough :)

jeromedockes committed:
better explaination of column capturing  intercept

jeromedockes committed:
errors_and_values -> solve

jeromedockes commented:
still TODO:
- make X actually sparse in tests
- remove separate `_errors` and `_values` pattern and do this logic in `fit`
- try to find a better name for `G_diag`

jeromedockes committed:
make test data design matrix actually sparse in ridgegcv tests

jeromedockes committed:
remove _errors_and_values

jeromedockes committed:
pep8

jeromedockes committed:
larger noise in tests

jeromedockes committed:
rename 'out' either squared_errors or predictions

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -1077,6 +1385,11 @@ def fit(self, X, y, sample_weight=None):
         -------
         self : object
         """
+        if np.any(self.alphas <= 0):
```
Usually, `X, y` are checked first.

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -1087,56 +1400,42 @@ def fit(self, X, y, sample_weight=None):
-            _pre_compute = self._pre_compute
-            _errors = self._errors
-            _values = self._values
+            decompose = self._decompose_gram
```
Naming `self._decompose_gram` and `self._decompose_covariance_*` is disconnected from `gcv_mode`, which controls which path to take. Maybe `self._decompose_eigen` and `self._decompose_svd_*` (and the same for `self._solve_covariance`.

ogrisel replied:
I would rather preserve the name of the matrix that is formed in the method name. Maybe we could be even more explicit and use:

```
self._svd_decompose_covariance
```
and:

```
self._eigen_decompose_gram
```

to know the size of the intermediate datastructures depending on `n_samples` and `n_features`.

jeromedockes replied:
I have made the change you suggest, and also changed the name of the method
which computes an svd of X by replacing "covariance" with "design_matrix".
so the names are now

- `_eigen_decompose_gram` (gcv_mode='eigen', eigen decomposition of gram matrix)
- `_svd_decompose_covariance` (gcv_mode='svd', eigen decomposition of covariance matrix, used when X is sparse)
- `_svd_decompose_design_matrix` (gcv_mode='svd', svd decomposition of design matrix, used when X is dense)

it is still not ideal because the second one is called `_svd...` and actually
computes an eigendecomposition, but this comes from the parameter choices used
to determine wether we solve the primal or dual problem, "svd" and "eigen"

ogrisel replied:
I had not realized the naming were so misleading.

Maybe we should actually rename those `gcv_mode` options to "gram and "covariance" to make it explicit which (while keeping backward compat with the old names with a deprecation warning).

BTW do you know if using a real svd on the design matrix would yield good results? Probably not for this PR but I am curious.

jeromedockes replied:
an SVD of the design matrix is indeed what we use when X is dense and n > p .
When X is sparse and n > p we don't want to actually store the left singular
vectors of X because they are a dense matrix of the same shape as X.
So we do something very similar but instead of writing the hat matrix as
(ignoring the intercept and sample weights for simplicity):

X = U S V^T
H = X(X^TX + alpha I)X^T = U (S^2 / (S^2 + alpha I)) U^T

we write it as

H = XV (1 / (S^2 + alpha I)) V^TX^T

and since we only need the dot product with y and the diagonal of this matrix 
we avoid creating a dense matrix of size (n x p).
(the diagonal is computed by batch in `_sparse_multidot_diag`)
So it is very similar to using an SVD of X, but in practice we only compute an
eigendecomposition of X^TX. We can consider that we are implicitly using an SVD
of X, writing U = XVS^-1 instead of explicitly forming U. This is being added
in this PR, as the sparse low-dimensional case was not handled before, so the
naming problem of choosing "svd" and computing an eigendecomposition does not
arise on master. (by the way, if you see a better way to compute Hy and diag(H)
when X is sparse I can change this implementation).

as you said, the name of the matrix that is being decomposed is more informative
than "svd" or "eigen", which can be deduced from whether the matrix is square or
not.

finally, I don't know why one would use another setting than "auto", besides
testing and checking that results are similar.

also note that all methods will result in memory problems if X is sparse and
large in both dimensions (n and p); in this case it is probably better not to
use generalized cross-validation or any decomposition of X.

thomasjpfan reviewed (commented):
This PR updates pretty often. 😅

jeromedockes committed:
Merge remote-tracking branch 'upstream/master' into ridge_gcv_sparse_x_intercept

jeromedockes commented on sklearn/linear_model/ridge.py:
```diff
@@ -921,6 +919,105 @@ def classes_(self):
 
+def _check_gcv_mode(X, gcv_mode):
+    possible_gcv_modes = [None, 'auto', 'svd', 'eigen']
+    if (gcv_mode is not None and not isinstance(gcv_mode, str)) or (
```
that was just in the odd chance that someone, confusing `gcv_mode` with `cv`, gave a numpy array for `gcv_mode`. I guess that's very improbable so I'll remove this

jeromedockes commented:
I haven't yet merged master because
`test_ridge_regression_dtype_stability[sparse_cg]` fails on master since
0dac63f48c5b3cd30e0a17dff5f5bbbc94bb4c39 (at least on my machine).

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -1001,18 +1079,164 @@ def _diag_dot(self, D, B):
+        X_mX = sqrt_sw[:, None] * safe_sparse_dot(
+            X_mean, X.T, dense_output=True)
+        X_mX_m = np.empty((n_samples, n_samples), dtype=X.dtype)
+        X_mX_m[:, :] = np.dot(X_mean, X_mean)
```
```python
        X_mX_m = np.full((n_samples, n_samples),
                         fill_value=np.dot(X_mean, X_mean), dtype=X.dtype)
```

ogrisel replied:
Actually, why not just:

```python
X_mX_m = np.dot(X_mean, X_mean)
```

or if dtype is not guaranteed to match:

```python
X_mX_m = np.dot(X_mean, X_mean).astype(X.dtype, copy=False)
```

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -927,6 +919,104 @@ def classes_(self):
+
+    def __init__(self, X, X_mean, sqrt_sw):
+        self.n_samples, self.n_features = X.shape
+        super().__init__(X.dtype, (self.n_samples, self.n_features + 1))
```
```python
n_samples, n_features = X.shape
super().__init__(X.dtype, (n_samples, n_features + 1))
```

thomasjpfan commented on sklearn/linear_model/ridge.py:
```diff
@@ -927,6 +919,104 @@ def classes_(self):
+
+    def __init__(self, X, X_mean, sqrt_sw):
+        self.n_samples, self.n_features = X.shape
+        super().__init__(X.dtype, (self.n_features + 1, self.n_samples))
```
```python
    def __init__(self, X, X_mean, sqrt_sw):
        n_samples, n_features = X.shape
        super().__init__(X.dtype, (n_features + 1, n_samples))
        ...

    def _matvec(self, v):
        n_features = self.shape[0]
        v = v.ravel()
        res = np.empty(n_features)
        ...

    def _matmat(self, v):
        n_features = self.shape[0]
        res = np.empty((n_features, v.shape[1]))
        ...
```

thomasjpfan reviewed (commented):
Another pass over the implementation.

jeromedockes committed:
address comments by @thomasjpfan

jeromedockes committed:
Merge remote-tracking branch 'upstream/master' into ridge_gcv_sparse_x_intercept

jeromedockes committed:
simplify _compute_gram

jeromedockes committed:
add tests for _compute_gram and _compute_covariance

jeromedockes committed:
assert np.allclose -> assert_allclose

jnothman commented:
This formally has 3 approvals. I've not looked at it. Is it ready for merge? Are we releasing in 0.21, or 0.22?

ogrisel commented:
I am still +1 for merge. It seems that all @thomasjpfan comments have been addressed but it would be great to have his confirmation before merging.

I think this is an important bug fix so +1 for 0.21 but I am fine to merge to master only for 0.22 if others disagree.

jeromedockes committed:
keep X in float64

jeromedockes committed:
more explicit decompose / solve private function names

now _{decomposition method used}_decompose_{decomposed matrix}
and _solve_{provided decomposition}_{matrix whose decomposition is provided}
e.g. _eigen_decompose_covariance, _solve_eigen_covariance

ogrisel commented:
We had an IRL discussion with @jeromedockes and we decided to stick to float64 for `RidgeCV` for now, otherwise we risk delaying this PR significantly.

This means that some of the changes from #13302 have been reverted. Only the `Ridge` class can natively work with `float32` data and even that seems to be problematic with the `sparse_cg` solver that seems to be unreliable.

jnothman commented:
@thomasjpfan please confirm and merge

thomasjpfan commented:
Looking into it. Going over the implementation one last time before merging.

thomasjpfan commented:
Thank you! @jeromedockes

agramfort commented:
Great job @jeromedockes 👏
